Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp

```diff
@@ -0,0 +1,226 @@
+        return nullptr;
+
+    // Compute all necessary transformations to build the tile image & the pattern.
+    FloatRect tileBoundaries;
+    AffineTransform tileImageTransform;
+    if (!buildTileImageTransform(renderer, *m_attributes, patternElement(), tileBoundaries, tileImageTransform))
+        return nullptr;
+
+    // Ignore 2D rotation, as it doesn't affect the size of the tile.
+    auto absoluteTransformIgnoringRotation = context.getCTM(GraphicsContext::DefinitelyIncludeDeviceScale);
+    SVGRenderingContext::clear2DRotation(absoluteTransformIgnoringRotation);
+    FloatRect absoluteTileBoundaries = absoluteTransformIgnoringRotation.mapRect(tileBoundaries);
+
+    // Scale the tile size to match the scale level of the patternTransform.
+    absoluteTileBoundaries.scale(static_cast<float>(m_attributes->patternTransform().xScale()), static_cast<float>(m_attributes->patternTransform().yScale()));
+
+    // Build tile image.
+    auto tileImage = createTileImage(*m_attributes, tileBoundaries, absoluteTileBoundaries, tileImageTransform);
+    if (!tileImage)
+        return nullptr;
+
+    auto tileImageSize = tileImage->logicalSize();
+
+    auto copiedImage = ImageBuffer::sinkIntoNativeImage(WTFMove(tileImage));
+    if (!copiedImage)
+        return nullptr;
+
+    // Compute pattern space transformation.
+    AffineTransform patternSpaceTransform;
+    patternSpaceTransform.translate(tileBoundaries.location());
+    patternSpaceTransform.scale(tileBoundaries.size() / tileImageSize);
+
+    AffineTransform patternTransform = m_attributes->patternTransform();
+    if (!patternTransform.isIdentity())
+        patternSpaceTransform = patternTransform * patternSpaceTransform;
+
+    // Build pattern.
+    return Pattern::create({ copiedImage.releaseNonNull() }, { true, true, patternSpaceTransform });
+}
+
+bool RenderSVGResourcePattern::buildPatternIfNeeded(GraphicsContext& context, const RenderLayerModelObject& targetRenderer)
+{
+    collectPatternAttributesIfNeeded();
+
+    m_pattern = buildPattern(context, targetRenderer);
+
+    if (!m_pattern)
+        return false;
+
+    return true;
+}
+
+bool RenderSVGResourcePattern::prepareFillOperation(GraphicsContext& context, const RenderLayerModelObject& targetRenderer, const RenderStyle& style)
+{
+    if (SVGHitTestCycleDetectionScope::isVisiting(*this))
+        return false;
+
+    if (!buildPatternIfNeeded(context, targetRenderer))
+        return false;
+
+    const auto& svgStyle = style.svgStyle();
+    context.setAlpha(svgStyle.fillOpacity());
+    context.setFillRule(svgStyle.fillRule());
+    context.setFillPattern(m_pattern.copyRef().releaseNonNull());
+    return true;
+}
+
+static AffineTransform transformOnNonScalingStroke(const RenderObject& object, const AffineTransform& resourceTransform)
+{
+    SVGGraphicsElement* element = downcast<SVGGraphicsElement>(object.node());
+    AffineTransform transform = element->getScreenCTM(SVGLocatable::DisallowStyleUpdate);
```

Comment:
Is that how RenderSVGShape computes the nonScalingStroke trafo? Should be unified